include/fftw-cpp/basic_plan.h

```diff
@@ -0,0 +1,462 @@
+        other.swap(*this);
+    }
+
+    template<size_t D, class Real, class Complex>
+    basic_plan_c2r<D, Real, Complex> &basic_plan_c2r<D, Real, Complex>::operator=(basic_plan_c2r &&other) noexcept {
+        basic_plan_c2r(std::move(other)).swap(*this);
+        return *this;
+    }
+
+    template<size_t D, class Real, class Complex>
+    void basic_plan_c2r<D, Real, Complex>::operator()() const {
+        fftw_execute(plan);
+    }
+
+    template<size_t D, class Real, class Complex>
+    template<typename ViewIn, typename ViewOut>
+    void basic_plan_c2r<D, Real, Complex>::operator()(ViewIn in, ViewOut out) const {
+        fftw_execute_dft_c2r(plan, detail::unwrap<false, Real, Complex>(in), detail::unwrap<true, Real, Complex>(out));
+    }
+
+    namespace detail {
+        // TODO currently we support one layout left and one layout right,
+        //  but that actually changes which dimension is which.
+        //  E. g. if input is 4x6 col-major and output is 6x4 row-major, now
+        //  the dimensions in the output are reversed, kind of defeating the purpose of layouts
+        //  If people really want to switch layouts they'll have to reverse it themselves.
+
+        template<size_t D>
+        std::array<int, D> dims(auto in, auto out) {
+            static_assert(D == 2u && "Currently only supporting 2D");
+            using layout_in = typename std::decay_t<decltype(in)>::layout_type;
+            using layout_out = typename std::decay_t<decltype(out)>::layout_type;
+
+            // TODO layouts other than right and left
+            constexpr bool in_right = std::same_as<MDSPAN::layout_right, layout_in>;
+            constexpr bool out_right = std::same_as<MDSPAN::layout_right, layout_out>;
+
+            auto Validate = [&](bool condition) {
+                if (!condition) throw std::invalid_argument("Extents don't match");
+            };
+
+            if (in_right == out_right) {
+                Validate(in.extent(0) == out.extent(0));
+                Validate(in.extent(1) == out.extent(1));
+            } else {
+                Validate(in.extent(0) == out.extent(1));
+                Validate(in.extent(1) == out.extent(0));
+            }
+
+            if constexpr (in_right) {
+                return {int(in.extent(0)), int(in.extent(1))};
+            } else {
+                return {int(in.extent(1)), int(in.extent(0))};
+            }
+
+        }
+
+        template<size_t ...I>
+        auto extents_impl(auto src, std::index_sequence<I...> indices) {
+            return std::array<int, sizeof...(I)>{int(src.extent(I))...};
+        }
+
+        template<size_t D>
+        std::array<int, D> extents(auto src) {
+            return extents_impl(src, std::make_index_sequence<D>());
+        }
+
+        template<size_t D>
+        std::array<int, D> dims_r2c(auto r, auto c) {
+            static_assert(D == 2u && "Currently only supporting 2D");
```

Comment:
I think this currently supports layout_right on one side and layout_left on the other. But a lot of this will get refactored anyway to support arbitrary strides